### ISSUE TITLE
Translation of Ralph's user interface into German

### ISSUE THREAD

**Issue by MuratKurnaz** [allegro/ralph#3602]
* Ralph version: 3
* Operating system: MacOS Big Sur Version 11.1
* Method of installation: Ralph development environment: https://ralph-ng.readthedocs.io/en/stable/development/development_environment/

Hello folks,

I'm currently working on Ralph 3 and trying to translate the interface into German. With a German django.po file, a few words can be translated. However, I am not able to translate some words like "warehouse" on the interface for adding a back-office asset (With the URL task end: back_office/backofficeasset/add/).

I have been searching the source code for weeks, but these words do not seem to be translatable, although I know it must be possible somehow. 

Can someone please help me.

Greetings 

Murat

Translated with www.DeepL.com/Translator (free version)

**Comment by NotJ3st3r** [comment]
I've been working on a german translation too. But it seems that the original .po is missing some strings. I've attached my german .po-file, but as you mentioned some translations are a bit odd.
[german.po.zip](https://github.com/allegro/ralph/files/5854647/german.po.zip)

**Comment by MuratKurnaz** [comment]
Yes, I have found that some translations are not realised by translating an English word with a German one. Variable extensions have to be done. I'll show you how I did it for the word warehouse. But first things first.

First you have to go to the file /src/ralph/settings/base.py. In the line (at least for me) is the variable LANGUAGE_CODE: with the value ("en-us") replace it with ("de-de"). 
OK, that was the first step. 

Now there are three types of translations (translations with the .po file excluded)

1. the translation of the tab. To translate a word (for example Warehouse, which is a sub-tab of Setting) in the tab go to the file src/ralph/admin/sitetrees.py Then go down to line 324 where you find 

section(_('Warehouse'), 'back_office', 'warehouse'), and translate Warehouse to (in my example) "Regionales Büro". Why Regionales Büro is not important. So now you shut down the programme and enter the command "make menu" so that the tab is regenerated and start the programme again with "make run". 

But that was only the translation of the tab, how do you translate the words in the rest of the Ralph system?

You do the other translations with verbose_name_(),. In the case of Warehouse, there is the class and the variable Warehouse. You go into the file src/admin/back_office/models.py and in line 58 you should find the following. 


class Warehouse( 
    AdminAbsoluteUrlMixin,
    NamedMixin,
    TimeStampMixin,
    models.Model 
):
         


(You extend this with)

class Warehouse(
    AdminAbsoluteUrlMixin,
    NamedMixin,
    TimeStampMixin,
    models.Model
):

     class Meta:
        verbose_name = _('Regionales Büro')
        verbose_name_plural = _('Regionale Büros')

Remember, we are dealing with Python, watch out for the indentations.

And now go to line 177 where you find the variable warehouse and extend it with verbose_name like this too:

    warehouse = models.ForeignKey(Warehouse, on_delete=models.PROTECT, verbose_name=_('Regionales Büro'))

I have noticed that most words are translated just by adding verbose_name to the respective variable. The words that also appear in the tab require additional translation for a complete translation (by complete, I mean that the word sons is not translated in other areas of Ralph):

     class Meta:
        verbose_name = _('x')
        verbose_name_plural = _('x')

You just have to find the respective variables of the words, but most of them are in the same Python scripts


3. Let's say you click on "Back Office" then on "Hardware" in the tab and want to add a "Back Office Asset" by clicking on "Add Back Office Asset" in the upper right corner. But think you want to translate that too. To translate such words you have to translate the English word into German in the respective HTML pages. How you find the words in the code is up to you, I use the search function of Visual Studio Code. In the case of "Add X", go to src/ralph/admin/templates/admin/change_list.html in line 41.

There you change "Add {{name}}" to "Add {{name}}".

So that's the way I translate Ralph.

Regards

Murat

**Comment by NotJ3st3r** [comment]
Then whats the purpose of the .po file? I thought all string that need to be translated will be stored in that file and only be linked in the source code to insert to right translation?

Wouldn't it be easyer to not hardcode the strings and instead use the .po file for every translation?

**Comment by MuratKurnaz** [comment]
I agree you and I dont know why the .po File doesnt work 100 %.

Some words are translatable through the .po File and some not. This is the way which work for me.
